### ISSUE TITLE
Why the perturbation with its value equals the maximum bound of Linf and L2 attack should be used in update? 

### ISSUE THREAD

**Issue by machanic** [max-andr/square-attack#5]
I wonder the use of maximum bound of Linf and L2 attack in the update is the best choice?
For example, I mean `deltas[i_img, :, center_h:center_h+s, center_w:center_w+s] = np.random.choice([-eps, eps], size=[c, 1, 1])`  in https://github.com/max-andr/square-attack/blob/master/attack.py#L236
Can it change to a value inside the `[-eps, eps]` interval? (smaller than `eps`) Would it possible be better?

**Comment by max-andr** [comment]
The whole idea of our attack is to avoid searching anything in the interior, i.e. (-eps, eps), and go directly to {-eps, eps} (which is also done, e.g., in https://arxiv.org/abs/1905.06635, https://arxiv.org/abs/1902.06894). I.e. to perform a search over the extreme points of the perturbation set.

Of course, you can try to deviate from this idea and see what comes out. But I wouldn't expect improved results. If you run, e.g., PGD for standard models, the final iterate will be almost always at an extreme point. So why bothering with searching for something in the interior?... :)

**Comment by machanic** [comment]
@max-andr I tested this algorithm, namely square attack. I found that it performs best in Linf norm attack. However, in ImageNet and TinyImageNet, the result of L2 norm attack is not the best. Prior-guided RGF (P-RGF) performs better than Square Attack. (Fewer queries). I have an idea that maybe we can combine random search and gradient estimation based method?

**Comment by max-andr** [comment]
Yeah, it totally makes sense that P-RGF achieves better results since it relies on a surrogate model while the Square Attack doesn't. We explicitly mentioned in the paper that we compare only attacks with the same knowledge model as we assume (i.e., no surrogate models). 

Of course, if one has a surrogate model, then one can fool many images with just 1 query and then run, e.g., the Square Attack on the rest of the images. Of course, this will very significantly decrease the average number of queries (my expectation: by a factor of 2 or 3 depending on the epsilon). But we always thought that having a surrogate model is already a very strong and somewhat questionable assumption since surrogate models are very expensive to obtain in general. Yes, it's easy to import from pytorch-vision lots of ImageNet models since image recognition on ImageNet is already a well-explored area. But what if one wants to fool, e.g., a text plagiarism system? It's totally unclear to me how to obtain not only a pre-trained model, but even a dataset for this task. 

> I have an idea that maybe we can combine random search and gradient estimation based method?

Yeah, sure, you can try this out. I can imagine that on *some* fraction of images a gradient estimation method may work better than the Square Attack.

**Comment by max-andr** [comment]
Closing this for now. Feel free to reopen if you have further questions or would like to have more explanations.
